### PR TITLE
Differentiate todo! and unimplemented!

### PR DESCRIPTION
**Pull request by llogiq**
This updates the panic message and docs to make it clear that `todo!` is for unfinished code and `unimplemented!` is for partial trait or enum impls.

r? @Centril

### CODE REVIEW

**Comment by Centril** [comment]
My understanding is that @dtolnay said (https://github.com/rust-lang/rust/pull/56348#issuecomment-447684679) we shouldn't differentiate like this.

r? @dtolnay cc @withoutboats @matklad

**Comment by matklad** [comment]
In practice people use `unimplemented!` for both "todo" and "trait signature forces me to write this method, but it doesn't make sense for my use case and is never called" (I saw the latter in rustc when trying to actively deprecate `unimplmented`). If we pick one of the meaning, some code will be "misusing" todo. 

I'd prefer to just say "unimplemented is an alias for `todo`", thus nudging folks towards using shorter and less ambiguous todo. Maybe even apply `#[doc(hidden)]` / `#[derecated]` to `unimplemented` in some distant future.

**Comment by petertodd** [comment]
Re: the partial trait use-case, here's an abbreviated example from one of my projects:

```
trait Zone {
    type Alllocator;
    fn allocator() -> Self::Allocator where Self: Default {
        todo!("Zone::allocator() required because {} implements Default", type_name::<Self>())
    }
}
```

This is a requirement for implementing the trait; it's a genuine `todo!()` as it's unfinished code. Indeed, you could imagine a more targeted [#[no_panic]](https://github.com/dtolnay/no-panic) actually check if any `todo!()`'s remained after release optimizations. Also, note the run-time error message:

```
thread 'main' panicked at 'not yet implemented: Zone::allocator() required because todotest::Heap implements Default', src/main.rs:7:9
```

The wording is something hasn't *yet* been implemented; it will be implemented in the future.

Conversely here's another example from the same project (again, very simplified):

```
trait Decode {
    type Error;
    fn decode(buf: &[u8]) -> Result<Self, Self::Error>;
}

impl Decode for ! {
    type Error = !;
    fn decode(buf: &[u8]) -> Result<Self, Self::Error> {
        panic!("! can't be decoded")
    }
}
```

At one point IIRC I actually used `unimplemented!()` for the above, but I removed it when I realized the error message was wrong: `decode()` will never be implemented for `!`, and the panicking implementation only exists at due to it being an edge case in my overall API design. (`unreachable!()` might be good here too)

tl;dr:

1. I really like how short `todo!()` is and use it all the time now.
2. I agree with @matklad on depreciating `unimplemented!()`

**Comment by withoutboats** [comment]
We've had this debate again and again during the development of todo, and my opinion is that std should not attempt to *prescribe* any particular way of using the APIs, because we all will use them differently.

Attempts to say that unimplemented is for a use case other than unfinished usually runs into the problem of differentiating unimplemented from unreachable.

*In my opinion,* panicking in a case that is neither unfinished nor unreachable (e.g. any public API which panics) should use `panic!` and contain a unique message explaining the error made in calling it. This leaves no use case for unimplemented in the way that I write Rust. However, I respect that other people have different ideas of what to use, and I think that's fine.

I don't understand the constant impulse to systematically prescribe every decision users make in writing code.

**Comment by withoutboats** [comment]
Just changing unimplemented's panic message to "not implemented" instead of "not *yet* implemented" seems fine to me, since it just makes it more flexible (though again, I would personally just use `panic` for anything I expected to merge to master). I just don't think we should put nearly as much energy as we already have into trying to dictate how these several semantically equivalent macros gets used.

**Comment by llogiq** [comment]
@withoutboats thank you! That's a good suggestion. I'll update the PR shortly.

File: src/libcore/macros/mod.rs [hunk 1]
```diff
@@ -556,13 +556,19 @@ macro_rules! unreachable {
...
-/// There is no difference between `unimplemented!` and `todo!` apart from the
-/// name.
+/// The difference between `unimplemented!` and [`todo!`](macro.todo.html) is that while `todo!`
+/// conveys an intent of implementing the functionality later and the message is "not yet
+/// implemented", `unimplemented!` makes no such claims. Its message is "not implemented".
+/// Also some IDEs will mark `todo!`s.
+///
+/// There will be cases where you need to partially implement a trait or implement a function
+/// for only some variants of an enum. In those cases, a plain [panic!](macro.panic.html) with a
+/// suitable message is advisable.
```
Comment (dtolnay) [inline src/libcore/macros/mod.rs]: I don't find this paragraph helpful, recommend removing it. Implementing some trait methods but not others, or implementing for some enum variants but not others, is exactly what most people would consider "not implemented".

If you want to draw a distinction against panic (which I do not think is necessary to do, and based on https://github.com/rust-lang/rust/pull/67445#issuecomment-568314669 neither does withoutboats) then it should be the distinction I voiced in https://github.com/rust-lang/rust/pull/56348#issuecomment-447684679: if there is a contract that the caller violated in calling the not-implemented code, then the salient point is that the caller violated the contract, not that the code is not implemented.

**Review by dtolnay — CHANGES_REQUESTED** [review]

File: src/libcore/macros/mod.rs (hunk 1 above)
Comment (llogiq) [reply src/libcore/macros/mod.rs]: I've removed the paragraph. I'll leave differentiating from `panic!` to a followup PR, if any.

**Review by dtolnay — APPROVED** [review]

File: src/libcore/macros/mod.rs [hunk 2]
```diff
@@ -605,19 +607,18 @@ macro_rules! unreachable {
...
+///         // at all.
+///         // This will display "thread 'main' panicked at 'not implemented'".
 ///         unimplemented!();
 ///     }
 ///
 ///     fn qux(&self) -> Result<u64, ()> {
-///         let n = self.bar();
 ///         // We have some logic here,
-///         // so we can use unimplemented! to display what we have so far.
+///         // We can add a message to unimplemented! to display our omission.
```
Comment (petertodd) [inline src/libcore/macros/mod.rs]: Minor typo/awkward wording. How about wording it as the following:

"We have some logic here. We can add a message to `unimplemented!` to explain our omission."

**Comment by petertodd** [comment]
@withoutboats That's fair! I didn't follow that discussion closely so I'm probably lacking context. :)
